### PR TITLE
Profile page of nonMembers only to be visible to SuperUsers.

### PR DESCRIPTION
**Pull request by heyrandhir**
### What is the change?
Issue: https://github.com/Real-Dev-Squad/website-members/issues/367

- Only Super Users should be able to view the Non-Members Profile page.

### Is it bug?
Yes

- Steps to repro
- Expected
- Actual

### \*Dev Tested?

- [x] Yes
- [ ] No

#### Platforms

- [x] Web

#### Browsers

- [x] Chrome
- [ ] Safari
- [ ] Firefox

### Before / After Change Screenshots

> For visual or interaction changes. Can be video / screenshot.

### CODE REVIEW

**Comment by harshith-venkatesh** [comment]
Why can't we think of making the call if and when it is undefined?
Instead of making call everytime....

**Comment by heyrandhir** [comment]
> Why can't we think of making the call if and when it is undefined? Instead of making call everytime....

Hi @harshith-venkatesh I have now added a check before making an API call. If the API is called previously it doesn't make the call again.

File: src/pages/[id]/index.js [hunk 1]
```diff
@@ -50,11 +47,26 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
...
+      id.charAt(0).toUpperCase() + id.slice(1)
+    } is not yet generated.`;
+    return <NotFound errorMsg={errorMsg} />;
+  }
   const { first_name = '', last_name = '' } = user;
   const memberName = `${first_name} ${last_name} | Member Real Dev Squad`;
 
-  const { query } = useRouter(); // this needs to be changed
-  const devUser = !!query.dev;
+  // const { query } = useRouter(); // this needs to be changed
```
Comment (prakashchoudhary07) [inline src/pages/[id]/index.js:68]: NIT: If not required we should just remove it

File: src/pages/[id]/index.js [hunk 2]
```diff
@@ -50,11 +47,26 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
     fillActiveTasksArray();
   }, [fillActiveTasksArray]);
 
+  if (errorMessage) {
+    return <NotFound errorMsg={errorMessage} />;
+  }
+
+  if (!isSuperUser) {
+    if (!apiCalledToVerifySuperUser) {
+      makeApiCallToVerifySuperUser();
```
Comment (prakashchoudhary07) [inline src/pages/[id]/index.js:56]: Since makeApiCallToVerifySuperUser is async shouldn't there be an await here?

Comment (heyrandhir) [reply src/pages/[id]/index.js:56]: I tried adding await here like this
```
    if (!apiCalledToVerifySuperUser) {
      (async () => await makeApiCallToVerifySuperUser())();
    }
``` 
but I am getting an error in the git log when committing as follows 
```
error  Redundant use of `await` on a return value  no-return-await
```

File: src/pages/[id]/index.js (hunk 1 above)
Comment (heyrandhir) [reply src/pages/[id]/index.js:68]: sure we can remove this.

**Review by prakashchoudhary07 — APPROVED** [review]

File: src/components/user-profile/index.js [hunk 3]
```diff
@@ -1,30 +1,23 @@
...
 const UserProfile = () => {
   const [isLoading, setIsLoading] = useState(true);
   const {
     user,
-    setUser,
     isSuperUser,
-    setIsSuperUser,
     isSuperUserMode,
     setIsSuperUserMode,
+    makeApiCallToVerifySuperUser,
```
Comment (ankushdharkar) [inline src/components/user-profile/index.js:12]: Nit: This can be a shorter verb "verifySuperUser". We don't need to write the imperative steps of _how_ we are doing it.
Otherwise, many functions will start looking like this:
- callFunctionFromClassWalletToCheckForAmountRemaining
- checkIfEmailIsValidByMatchingRegularExpressionForEmail

File: src/components/user-profile/index.js [hunk 4]
```diff
@@ -1,30 +1,23 @@
...
       setIsLoading(false);
-      setUser(data);
-      if (data) {
-        setIsSuperUser(data.roles.super_user);
-      }
-    }
-    getUserProfile();
-  }, [setUser, setIsSuperUser]);
+    })();
+  }, [makeApiCallToVerifySuperUser]);
```
Comment (ankushdharkar) [inline src/components/user-profile/index.js:20]: API Call in useEffect :O

File: src/pages/[id]/index.js [hunk 5]
```diff
@@ -50,11 +47,25 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
     fillActiveTasksArray();
   }, [fillActiveTasksArray]);
 
+  if (errorMessage) {
```
Comment (ankushdharkar) [inline src/pages/[id]/index.js:50]: Nice!

File: src/pages/[id]/index.js [hunk 6]
```diff
@@ -50,11 +47,25 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
...
+    return <NotFound errorMsg={errorMessage} />;
+  }
+
+  if (!isSuperUser) {
+    if (!apiCalledToVerifySuperUser) {
```
Comment (ankushdharkar) [inline src/pages/[id]/index.js:55]: Nit: Could have been just one If here.

File: src/pages/[id]/index.js [hunk 7]
```diff
@@ -50,11 +47,25 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
...
+      makeApiCallToVerifySuperUser();
+    }
+  }
+  if (!isSuperUser && !user.roles?.member) {
+    const errorMsg = `The Member Page for ${
+      id.charAt(0).toUpperCase() + id.slice(1)
```
Comment (ankushdharkar) [inline src/pages/[id]/index.js:61]: This seems a little out of the ordinary.
Why not use the first and last names here, and show userId as it is?

File: src/store/user/user-context.js [hunk 8]
```diff
@@ -10,6 +11,18 @@ export const UserContextProvider = ({ children }) => {
   const [showMemberRoleUpdateModal, setShowMemberRoleUpdateModal] =
     useState(false);
   const [selectedMember, setSelectedMember] = useState(null);
+  const [apiCalledToVerifySuperUser, setApiCalledToVerifySuperUser] =
```
Comment (ankushdharkar) [inline src/store/user/user-context.js:14]: Could we please use simpler names here, as described here https://github.com/Real-Dev-Squad/website-members/pull/410/files#r952055322?

File: src/store/user/user-context.js [hunk 9]
```diff
@@ -10,6 +11,18 @@ export const UserContextProvider = ({ children }) => {
...
+    useState(false);
+
+  const makeApiCallToVerifySuperUser = async () => {
+    const { data } = await getUserSelf();
```
Comment (ankushdharkar) [inline src/store/user/user-context.js:18]: What about error handling? What happens if the request encounters a 4xx or 5xx?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/store/user/user-context.js (hunk 8 above)
Comment (heyrandhir) [reply src/store/user/user-context.js]: yeah right. Updated the variable names.

File: src/pages/[id]/index.js (hunk 7 above)
Comment (heyrandhir) [reply src/pages/[id]/index.js]: Yeah, I have updated the code now to use the first name & last name.

File: src/pages/[id]/index.js (hunk 6 above)
Comment (heyrandhir) [reply src/pages/[id]/index.js]: merged both the conditions into one. Thanks.

File: src/components/user-profile/index.js (hunk 3 above)
Comment (heyrandhir) [reply src/components/user-profile/index.js]: Yeah Right. Thanks 🙂

File: src/pages/[id]/index.js (hunk 5 above)
Comment (heyrandhir) [reply src/pages/[id]/index.js:46]: Thanks 😇

File: src/components/user-profile/index.js (hunk 4 above)
Comment (heyrandhir) [reply src/components/user-profile/index.js]: I need to read more about this. I will create a new issue and work on this part. Resolving this so that it can be shipped to production and used.

File: src/store/user/user-context.js [hunk 10]
```diff
@@ -10,6 +11,24 @@ export const UserContextProvider = ({ children }) => {
...
+  const verifySuperUser = async () => {
+    setUser(null);
+    setIsSuperUser(false);
+    setSuperUserVerified(false);
+    try {
+      const { data } = await getUserSelf();
+      if (data) {
+        setUser(data);
+        setIsSuperUser(Boolean(data.roles?.super_user));
+        setSuperUserVerified(true);
```
Comment (ivinayakg) [inline src/store/user/user-context.js:25]: if both the `isSuperUser` and `superUserVerified` are getting settled to true at the same time then can't we just switch to one state instead of multiple cause I found it redundant?

Comment (heyrandhir) [reply src/store/user/user-context.js:25]: `isSuperUser` checks if a user is or is not a superUser but `superUserVerified` checks if the API call to verify super User is completed. So the next time we don't make an API call again to verify whether or not the user is a superUser.

File: src/store/user/user-context.js [hunk 11]
```diff
@@ -10,6 +11,27 @@ export const UserContextProvider = ({ children }) => {
...
+    let selfSuperUserVerified = false;
+    try {
+      const { data } = await getUserSelf();
+      if (data) {
+        selfUser = data;
+        selfIsSuperUser = Boolean(data.roles?.super_user);
+        selfSuperUserVerified = true;
+      }
+    } catch {
+      // by default handled in the initial local variables
```
Comment (ankushdharkar) [inline src/store/user/user-context.js:28]: Nit: console.error here would have been nice

File: src/store/user/user-context.js [hunk 12]
```diff
@@ -10,6 +11,27 @@ export const UserContextProvider = ({ children }) => {
...
+  const verifySuperUser = async () => {
+    let selfUser = null;
+    let selfIsSuperUser = false;
+    let selfSuperUserVerified = false;
+    try {
+      const { data } = await getUserSelf();
+      if (data) {
+        selfUser = data;
+        selfIsSuperUser = Boolean(data.roles?.super_user);
+        selfSuperUserVerified = true;
```
Comment (ankushdharkar) [inline src/store/user/user-context.js:25]: Why is this needed? Why can't we just reuse `selfIsSuperUser`?

File: src/components/user-profile/index.js [hunk 13]
```diff
@@ -1,30 +1,23 @@
...
       setIsLoading(false);
-      setUser(data);
-      if (data) {
-        setIsSuperUser(data.roles.super_user);
-      }
-    }
-    getUserProfile();
-  }, [setUser, setIsSuperUser]);
+    })();
+  }, [verifySuperUser]);
```
Comment (ankushdharkar) [inline src/components/user-profile/index.js:20]: This should ideally not be the responsibility of the component. It should exist in an API util, or data layer.

File: src/pages/[id]/index.js [hunk 14]
```diff
@@ -50,11 +43,21 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
     fillActiveTasksArray();
   }, [fillActiveTasksArray]);
 
+  if (errorMessage) {
+    return <NotFound errorMsg={errorMessage} />;
+  }
+
+  if (!isSuperUser && !superUserVerified) {
```
Comment (ankushdharkar) [inline src/pages/[id]/index.js:50]: `superUserVerified` checks for whether the call was made or not? From the name, it seems like it verifies if the current user is indeed a super user.

File: src/store/user/user-context.js [hunk 15]
```diff
@@ -10,6 +11,27 @@ export const UserContextProvider = ({ children }) => {
...
     useState(false);
   const [selectedMember, setSelectedMember] = useState(null);
+  const [superUserVerified, setSuperUserVerified] = useState(false);
+
+  const verifySuperUser = async () => {
+    let selfUser = null;
+    let selfIsSuperUser = false;
+    let selfSuperUserVerified = false;
+    try {
+      const { data } = await getUserSelf();
```
Comment (ankushdharkar) [inline src/store/user/user-context.js:21]: If the API fails/errors, will we never set `selfSuperUserVerified`, potentially causing infinite loops in consuming components e.g [here](https://github.com/Real-Dev-Squad/website-members/pull/410/files#r953219362)?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/store/user/user-context.js (hunk 12 above)
Comment (heyrandhir) [reply src/store/user/user-context.js]: probably the name 'selfSuperUserVerified`confused you and vinayak-g. I have changed it to `selfUserApiCalled`. This variable prevents us from making multiple API calls.

File: src/components/user-profile/index.js (hunk 13 above)
Comment (heyrandhir) [reply src/components/user-profile/index.js]: Agreed. I will create a new ticket to move this into separate API util or data layer.

File: src/pages/[id]/index.js (hunk 14 above)
Comment (heyrandhir) [reply src/pages/[id]/index.js]: yeah, the name `superUserVerified ` was confusing. renamed it to `userApiCalled`

File: src/store/user/user-context.js (hunk 15 above)
Comment (heyrandhir) [reply src/store/user/user-context.js:21]: even if the API fails/errors after the try-catch block we call `setUserApiCalled` which will set it to `false`( the initial value we initialized `selfUserApiCalled` with).

**Comment by heyrandhir** [comment]
the isOption in the code will be fetched from context. it will be taken care in the refactor PR. https://github.com/Real-Dev-Squad/website-members/issues/420

File: src/pages/[id]/index.js [hunk 16]
```diff
@@ -50,11 +43,21 @@ const MemberProfile = ({ imageLink, user, contributions, errorMessage }) => {
     fillActiveTasksArray();
   }, [fillActiveTasksArray]);
 
+  if (errorMessage) {
+    return <NotFound errorMsg={errorMessage} />;
+  }
+
+  if (!isSuperUser && !userApiCalled) {
+    setUserPrivileges();
+  }
```
Comment (heyrandhir) [inline src/pages/[id]/index.js:52]: This is a potential threat that can cause the page to infinitely re-render. I have created a ticket for this https://github.com/Real-Dev-Squad/website-members/issues/419. Will be refactoring it there.
